internal/controller/services/monitoring/monitoring_controller.go

```diff
@@ -98,3 +108,129 @@ func (h *serviceHandler) NewReconciler(ctx context.Context, mgr ctrl.Manager) er
+	if tracingSpec == nil {
+		// Ensure Tempo instance is absent
+		return removeTempoInstance(ctx, rr, namespace)
+	}
+	if tracingSpec.Storage.Backend == "" {
+		tracingSpec.Storage.Backend = "pv"
+	}
+	// Deploy the appropriate Tempo instance based on the backend
+	if tracingSpec.Storage.Backend == "pv" {
+		return deployTempoMonolithic(ctx, rr, tracingSpec, namespace)
+	}
+	return deployTempoStack(ctx, rr, tracingSpec, namespace)
+}
+
+func deployTempoMonolithic(ctx context.Context, rr *types.ReconciliationRequest, tracingSpec *serviceApi.TracesSpec, namespace string) error {
+	tempo := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "tempo.grafana.com/v1alpha1",
+			"kind":       "TempoMonolithic",
+			"metadata": map[string]interface{}{
+				"name":      "tempo",
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"multitenancy": map[string]interface{}{
+					"enabled": true, // Required for OpenShift
+				},
+				"storage": map[string]interface{}{
+					"traces": map[string]interface{}{
+						"backend": "pv",
+						"size":    tracingSpec.Storage.Size, // Optional
+					},
+				},
+			},
+		},
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, rr.Client, tempo, func() error {
+		return nil
+	}); err != nil {
+		return errors.New("failed to deploy TempoMonolithic")
+	}
+	return nil
+}
+
+func deployTempoStack(ctx context.Context, rr *types.ReconciliationRequest, tracingSpec *serviceApi.TracesSpec, namespace string) error {
+	// Validate required fields for TempoStack
+	if tracingSpec.Storage.Secret == "" {
+		return errors.New("secret name is required when backend is not 'pv'")
+	}
+
+	if tracingSpec.Storage.Size != "" {
+		return errors.New("size is not supported in TempoStack CRD")
+	}
+
+	tempo := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "tempo.grafana.com/v1alpha1",
+			"kind":       "TempoStack",
+			"metadata": map[string]interface{}{
+				"name":      "tempo",
+				"namespace": namespace,
+			},
+			"spec": map[string]interface{}{
+				"template": map[string]interface{}{
+					"gateway": map[string]interface{}{
+						"enabled": true, // Required for OpenShift mode
+					},
+				},
+				"multitenancy": map[string]interface{}{
+					"enabled": true, // Required for OpenShift
+				},
+				"tenants": map[string]interface{}{
+					"mode": "openshift",
+				},
+				"storage": map[string]interface{}{
+					"secret": map[string]interface{}{
+						"name": tracingSpec.Storage.Secret,
+						"type": tracingSpec.Storage.Backend,
+					},
+				},
+			},
+		},
+	}
+
+	if _, err := controllerutil.CreateOrUpdate(ctx, rr.Client, tempo, func() error {
+		return nil
+	}); err != nil {
+		return errors.New("failed to deploy TempoStack")
+	}
+	return nil
+}
+
+func removeTempoInstance(ctx context.Context, rr *types.ReconciliationRequest, namespace string) error {
```

Comment:
you may want to add a `gc` action to the controller, so it would automatically delete outdated resources when the spec changes.

Reply:
have oyu tried with the gc action ?